django/db/models/query.py

```diff
@@ -890,31 +890,73 @@ def bulk_update(self, objs, fields, batch_size=None):
         max_batch_size = connection.ops.bulk_batch_size(["pk", "pk"] + fields, objs)
         batch_size = min(batch_size, max_batch_size) if batch_size else max_batch_size
         requires_casting = connection.features.requires_casted_case_in_updates
+        requires_casting_in_values = (
+            connection.features.requires_casted_case_in_values_updates
+        )
         batches = (objs[i : i + batch_size] for i in range(0, len(objs), batch_size))
         updates = []
+
+        has_related_fields = any(f.model is not self.model for f in fields)
+        has_field_references = False
         for batch_objs in batches:
-            update_kwargs = {}
-            for field in fields:
-                when_statements = []
-                for obj in batch_objs:
+            row_tuples = []
+            for obj in batch_objs:
+                values = [Value(getattr(obj, "pk"))]
+                for field in fields:
                     attr = getattr(obj, field.attname)
-                    if not hasattr(attr, "resolve_expression"):
+                    if hasattr(attr, "resolve_expression"):
+                        has_field_references = True
+                    else:
                         attr = Value(attr, output_field=field)
-                    when_statements.append(When(pk=obj.pk, then=attr))
-                case_statement = Case(*when_statements, output_field=field)
-                if requires_casting:
-                    case_statement = Cast(case_statement, output_field=field)
-                update_kwargs[field.attname] = case_statement
-            updates.append(([obj.pk for obj in batch_objs], update_kwargs))
+
+                    if requires_casting_in_values:
+                        attr = Cast(attr, output_field=field)
+                    values.append(attr)
+
+                row_tuples.append(RowTuple(*values))
+            updates.append(
+                RowTupleValues(
+                    row_tuples, pk_field=self.model._meta.pk, field_list=fields
```

Comment:
In a world where we have `CompositeField` this could be changed to `output_field=CompositeField(fields=fields)`.